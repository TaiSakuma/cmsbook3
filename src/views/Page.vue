<template>
  <div class="page">
    <v-breadcrumbs :items="breadcrumbsItems"></v-breadcrumbs>
    <v-card outlined max-width="980" class="ma-3 pa-3">
      <v-card-text class="markdown-body" v-html="content"></v-card-text>
    </v-card>
  </div>
</template>

<script>
import marked from "marked";

// import Prism from 'prism-es6'; // Jest doesn't work with this
import Prism from 'prismjs';
import "prism-es6/components/prism-markup-templating";
import "prism-es6/components/prism-python";
import "prism-es6/components/prism-bash";
import "prism-es6/components/prism-latex";
import "@/prism.css";

import { retrieveFrom } from "@/cmsbook3-retrieve";

export default {
  name: "page",
  data() {
    return {
      cmsbook_url: process.env.VUE_APP_CMSBOOK_URL,
      md: "",
    };
  },
  computed: {
    content() {
      let htmlString;
      try {
        htmlString = marked(this.md);
      } catch (error) {
        return "cannot parsed as markdown";
      }

      try {
        return this.editHtml(htmlString);
      } catch (error) {
        console.error(error);
        return htmlString;
      }
    },
    path() {
      return `/${this.$route.params.chapter}/${this.$route.params.section}/${this.$route.params.page}`;
    },
    breadcrumbsItems() {
      let ret = [{ text: this.$route.params.chapter }];
      if (this.$route.params.section) {
        ret.push({ text: this.$route.params.section });
      }
      if (this.$route.params.page) {
        this.$route.params.page.split("/").forEach((e) => {
          ret.push({ text: e });
        });
      }
      return ret;
    },
  },
  updated: function () {
    Prism.highlightAll();
    try {
      MathJax.Hub.Typeset();
    } catch (error) {
      console.log(error);
    }
  },
  watch: {
    path: {
      handler: function () {
        this.getMarkDownFromPath();
      },
      immediate: true,
    },
  },
  methods: {
    async getMarkDownFromPath() {
      if (!this.path) {
        this.md = "";
        return;
      }
      try {
        this.md = await retrieveFrom(this.path);
      } catch (error) {
        this.md = "Error: cannot get: " + this.path;
      }
    },
    editHtml(htmlString) {
      const pathToCurrentDir = this.cmsbook_url + this.$route.path.match(/.*\//);
      let ret = this.editHtmlWithJQuery(htmlString, pathToCurrentDir);
      return ret;
    },
    editHtmlWithJQuery(htmlString, pathToCurrentDir) {
      const $ = require( "jquery" );

      let tree = $(`<div>${htmlString}</div>`);
      tree.find("a:not([href^='#'])").attr("target", "_blank");

      tree
        .find(
          "a:not([href^='http:'],[href^='https:'],[href^='file:'],[href^='/'],[href^='#'])"
        )
        .each(function () {
          this.setAttribute(
            "href",
            this.getAttribute("href").replace(/^/, pathToCurrentDir)
          );
        });

      tree
        .find("img:not([src^='http:'],[src^='https:'],[src^='/'])")
        .each(function () {
          this.setAttribute("src", this.getAttribute("src").replace(/^/, pathToCurrentDir));
        });

      return tree.html();
    },
  },
};
</script>

<!-- https://github.com/sindresorhus/github-markdown-css -->
<style>
@import "../../node_modules/github-markdown-css/github-markdown.css";

.markdown-body {
  /* box-sizing: border-box; */
  min-width: 200px;
  /* max-width: 980px; */
  /* border-radius: 10px; */
}

.v-application .markdown-body code {
  box-shadow: none;
  color: inherit;
}
.v-application .markdown-body code:before {
  content: "";
}

.theme--dark .markdown-body {
  color: #f5f5f5;
}

.theme--dark .markdown-body hr {
  background-color: #757575;
}

.theme--dark .markdown-body blockquote {
  border-left-color: #757575;
  color: #bdbdbd;
}

.theme--dark .markdown-body pre {
  background-color: #212121;
}

.theme--dark .markdown-body code {
  background-color: inherit;
  /* color: #f44336; */
}

.theme--light .markdown-body code {
  background-color: inherit;
}

.theme--light .markdown-body pre {
  background-color: #f6f8fa;
}

.theme--dark .markdown-body table td,
.theme--dark .markdown-body table th {
  border: 1px solid #616161;
}

.theme--dark .markdown-body table tr {
  background-color: #212121;
}

.theme--dark .markdown-body table tr:nth-child(2n) {
  background-color: #424242;
}
</style>
