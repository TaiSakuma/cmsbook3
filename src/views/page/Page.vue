<template>
  <div class="page">
    <v-breadcrumbs :items="breadcrumbsItems"></v-breadcrumbs>
    <div outlined max-width="980" class="ma-3 pa-3">
      <div class="markdown-body" v-html="content"></div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { computed } from "vue";
import { useRoute } from "vue-router";

import { useContent } from "./content";
import { usePrism } from "./prism";
import { useMathJax } from "./mathjax";

const route = useRoute();

const { content } = useContent();

const breadcrumbsItems = computed(() => {
  let ret = [{ title: route.params.chapter }];
  if (route.params.section) {
    ret.push({ title: route.params.section });
  }
  if (Array.isArray(route.params.page)) {
    route.params.page.forEach((e) => {
      ret.push({ title: e });
    });
  } else if (route.params.page) {
    ret.push({ title: route.params.page });
  }
  return ret;
});

usePrism();
useMathJax();
</script>

<style>
/* .v-application .markdown-body code:before {
  content: "";
} */
</style>
