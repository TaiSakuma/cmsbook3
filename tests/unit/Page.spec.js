import Vue from "vue";
import VueRouter from "vue-router";
import Vuetify from "vuetify";
import { mount, shallowMount, createLocalVue } from "@vue/test-utils";

import moxios from "moxios";

import $ from "jquery";

import Page from "@/views/Page.vue";

Vue.use(Vuetify);
// Vue.use(VueRouter);

describe("Page.vue", () => {
  const ENV_ORG = process.env;
  const $_ORG = global.$;
  const MathJax_ORG = global.MathJax;

  let localVue;
  let router;
  let wrapper;
  beforeEach(() => {
    process.env.VUE_APP_CMSBOOK_URL = "http://localhost/cmsbook";
    global.$ = $;
    global.MathJax = { typeset: jest.fn() };
    moxios.install();
    localVue = createLocalVue();
    router = new VueRouter();

    wrapper = mount(Page, {
      localVue,
      router,
      mocks: {
        $route: {
          path: "/chapter-A/section-b/page-3.md",
          params: {
            chapter: "chapter-A",
            section: "section-b",
            page: "page-3.md"
          }
        }
      }
    });
  });

  afterEach(() => {
    moxios.uninstall();
    global.MathJax = MathJax_ORG;
    global.$ = $_ORG;
    process.env = ENV_ORG;
  });

  it("server URL", () => {
    expect(wrapper.vm.path).toBe(
      "http://localhost/cmsbook/chapter-A/section-b/page-3.md"
    );
  });

  it("axios request ", done => {
    moxios.wait(() => {
      let request = moxios.requests.mostRecent();
      expect(request.config.url).toBe(
        "http://localhost/cmsbook/chapter-A/section-b/page-3.md"
      );
      expect(request.config.method).toBe("get");
      expect(request.config.data).toBeUndefined();
      done();
    });
  });

  it("snapshot", done => {
    moxios.wait(async () => {
      let request = moxios.requests.mostRecent();
      await request.respondWith({
        status: 200,
        response: "**marked**"
      });
      expect(wrapper.html()).toMatchSnapshot();
      done();
    });
  });

  it("marked", done => {
    console.log(wrapper.vm.path);
    moxios.wait(async () => {
      let request = moxios.requests.mostRecent();
      await request.respondWith({
        status: 200,
        response: "**marked**"
      });
      expect(wrapper.html()).toContain("<strong>marked</strong>");
      done();
    });
  });

  it("404", done => {
    moxios.wait(async () => {
      let request = moxios.requests.mostRecent();
      await request.respondWith({
        status: 404
      });
      expect(wrapper.text()).toContain("Error: cannot get");
      done();
    });
  });

  it("test text", done => {
    console.log(wrapper.vm.path);
    moxios.wait(async () => {
      let request = moxios.requests.mostRecent();

      await request.respondWith({
        status: 200,
        response: "**marked**"
      });

      console.log(request.config);
      console.log(wrapper.html());
      done();
    });
  });
});
