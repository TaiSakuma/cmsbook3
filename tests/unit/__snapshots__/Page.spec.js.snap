// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page.vue snapshot 1`] = `
<div class="page">
  <ul class="v-breadcrumbs theme--light">
    <li>
      <div class="v-breadcrumbs__item">chapter-A</div>
    </li>
    <li class="v-breadcrumbs__divider">/</li>
    <li>
      <div class="v-breadcrumbs__item">section-b</div>
    </li>
    <li class="v-breadcrumbs__divider">/</li>
    <li>
      <div class="v-breadcrumbs__item">page-3.md</div>
    </li>
  </ul>
  <div class="container pt-0 container--fluid">
    <div class="row">
      <div class="markdown-body mx-5 pa-5 col">
        <p><strong>marked</strong></p>
      </div>
    </div>
  </div>
</div>
`;
